src/app/server/Server.cpp

```diff
@@ -415,4 +413,43 @@ CHIP_ERROR Server::AddTestCommissioning()
     return err;
 }
 
+static void ServerScheduleInit(intptr_t context)
+{
+    AppDelegate * delegate = reinterpret_cast<AppDelegate *>(context);
+    Server::GetInstance().Init(delegate, CHIP_PORT, CHIP_UDC_PORT, Inet::InterfaceId::Null());
+}
+
+CHIP_ERROR MatterServerInit(AppDelegate * delegate, uint16_t secureServicePort, uint16_t unsecureServicePort,
+                            Inet::InterfaceId interfaceId)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    err = chip::Platform::MemoryInit();
+    SuccessOrExit(err);
+
+    err = DeviceLayer::PlatformMgr().InitChipStack();
+    SuccessOrExit(err);
+
+    err = Server::GetInstance().Init(delegate, secureServicePort, unsecureServicePort, interfaceId);
+
+exit:
+    return err;
+}
+
+CHIP_ERROR MatterServerScheduleInit(AppDelegate * delegate)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    err = chip::Platform::MemoryInit();
+    SuccessOrExit(err);
+
+    err = DeviceLayer::PlatformMgr().InitChipStack();
+    SuccessOrExit(err);
+
+    chip::DeviceLayer::PlatformMgr().ScheduleWork(ServerScheduleInit, reinterpret_cast<intptr_t>(delegate));
```

Comment:
I have the problem understanding why forcing all applications to follow this initialization sequence helps. Could you help me understand?

Reply:
The problem is the Matter*PluginServerInitCallback functions are currently implemented as if they are called once per process lifetime, which register delegates to platform, but CHIP stack init(chip::DeviceLayer::PlatformMgr().InitChipStack()) is handled in the application logic.
1. For server example, If the CHIP stack is shut down and restarted (e.g. via PlatformManager::Shutdown and then a new InitChipStack, etc), those registered delegates might not valid.
2. For client example, any time DeviceController::Init is called (which can happen during process lifetime due to shutting down an existing DeviceController and starting a new one), it will call InitDataModelHandler, which will call these callback functions try and register their attribute access callbacks.

Reply:
If we force the initialization sequence, and hide all details to app, it can solve the various race condition during initialization and solve the two issues above by making cluster layer and platform layer always init and shutdown together

Reply:
> If we force the initialization sequence, and hide all details to app, it can solve the various race condition during initialization and solve the two issues above by making cluster layer and platform layer always init and shutdown together

Hiding does not solve the problem. The real problem is that the public APIs are hard to use properly and there are too many bad assumptions there. If implementers/integrators shoot themselves in the foot during init, we can address some of it at the root.

I will admit that already `Server` class as a whole is this general idea of "These are all the bits and pieces to be a Matter device with cluster servers" with all the necessary inits. We can continue incrementally reworking the APIs to be beter.

Overall, we do have a huge problem that `Matter*PluginServerInitCallback` are all global and argument-less and called by generated code. This prevents orderly init of just what you care about, and has caused many developers to use shortcuts where they make use of many global variables in cluster server code just to deal with the fact that there is no clean way to pass init argument to the cluster inits...

@Damian-Nordic @bzbarsky-apple @mrjerryjohns